Add emerge hooks

Schievel1 opened the pull request:
Right now Portage has already bashrc to hook into the ebuild process. However, initially this was thought to manipulate environment variables (I think so, hence the name?) for a single package build.
Bashrc is sourced several times DURING the ebuild process. However it is not sourced BEFORE and AFTER the ebuild process.

My proposed feature would be to source emergerc, a script located in /etc/portage as well, several times and change a environment variable (e.g. EMERGE_PHASE) to let the users control which code is called.
A purpose for that for example would be to make a btrfs snapshot every time before emerging an update.

There are several situations that come to mind, where a user could benefit from a source of emergerc. The value that the environment variable $EMERGE_PHASE will have it that situation is written behind those points:
- every time emerge starts (emerge_startup)
- right before the dependencies are calculated (pre_calc_deps)
- after dependencies are successfully calculated (post_calc_deps_success)
- after dependencies and they were not successfully calculated (post_calc_deps_fail)
- after the prompt "Would you like to merge these packages?" (if --ask) right before emerge starts working with the first package (pre_first_emerge)
- before --sync (pre_sync)
- after --sync (although there is postsync.d for that) (post_sync_success) / (post_sync_fail)
- when emerge installed all packages successfully  (post_emerge_success)
- when emerge failed with one package (post_emerge_fail)
- when emerge exits (emerge_exit)
- before and after depclean (pre_depclean) / (post_depclean)
- before and after unmerge (pre_unmerge) / (post_unmerge)

To tell the different phases of emerge apart, one would do the same like in bashrc with environment variables
https://wiki.gentoo.org/wiki//etc/portage/bashrc#Environment_variable
```
if [ "${EMERGE_PHASE}" == "pre_first_emerge" ]
then
  echo "your code here"
fi
```

See also:
Bug: https://bugs.gentoo.org/839777

Schievel1 committed:
added function to call /etc/portage/emergerc

And several call:
- every time emerge starts (emerge_startup)
- right before the dependencies are calculated (pre_calc_deps)
- after dependencies are successfully calculated (post_calc_deps_success)
- after dependencies and they were not successfully calculated (post_calc_deps_fail)
- after the prompt "Would you like to merge these packages?" (if --ask) right before emerge starts working with the first package (pre_first_emerge)
- before --sync (pre_sync)
- after --sync (although there is postsync.d for that) (post_sync_success) / (post_sync_fail)
- when emerge installed all packages successfully  (post_emerge_success)
- when emerge failed with one package (post_emerge_fail)
- when emerge exits (emerge_exit)
- before and after depclean (pre_depclean) / (post_depclean)
- before and after unmerge (pre_unmerge) / (post_depclean)

Closes: https://bugs.gentoo.org/839777
Closes: https://bugs.gentoo.org/272988

Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>

Schievel1 commented:
This works, I have no idea whatsoever about the security implications when calling a shell script like that in python. 
Maybe someone has some better suggestions and I would be happy to put hat in.

Schievel1 commented:
To check out when which phases are run, use 
```
#!/bin/bash

echo ${EMERGE_PHASE}
```
in `/etc/portage/emergerc`

floppym commented on lib/_emerge/actions.py:
```diff
@@ -99,6 +99,10 @@
 
+emergerc_script = os.path.join("/", portage.const.USER_CONFIG_PATH, "emergerc")
+def callemergerc(phase):
+    os.system(" [[ -f " + emergerc_script + " ]] && export EMERGE_PHASE=" + phase + " && " + os.path.join(". /", emergerc_script))
```
Don't use `os.system` for this, especially if `emergerc` is expected to be a bash script (not POSIX-compatible sh).

The `portage.process.spawn_bash` function might be more appropriate if `emergerc` is to be sourced as a bash script.

Alternatively, we could require that `emergerc` be executable and call `portage.process.spawn` or `subprocess.run` on it directly.

The `EMERGE_PHASE` environment var could be passed in via the `env` keyword argument.

Instead of testing for the existence of `emergerc_script` in bash, you could do it in python directly using `os.path.isfile`.

Schievel1 replied:
Hi, thanks for you suggestions.

That's a lot of options, and I don't really understand the pros and cons for them. Which one would you prefer the most? 

I will do the isFile check in python, no problemo 👌

floppym replied:
I think I would opt for `subprocess.run`, since we don't really have a need for the more advanced features of `portage.process.spawn`.

Schievel1 committed:
function to call emergerc is now subprocess.run

instead of using os.system()

Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>

Schievel1 commented:
Do you have any ideas what else could be passed as an environment variable?

Schievel1 committed:
send packages and options to hook where possible

Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>

Schievel1 commented:
I added some functionality to send the list of packages and the command line options of emerge to the hook, so users can write scripts that react to that accordingly. I converted this back to draft for the time being, because I still need to test that thoroughly.

Schievel1 committed:
send the PID of the current emerge process to emergerc

Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>

MrByte commented:
Noob question. Can you make an ebuild for those of us that want to use it, ala local overlay?

Schievel1 commented:
> Noob question. Can you make an ebuild for those of us that want to use it, ala local overlay?

Not really. You can make a user patch, but the script is too much of a moving target. I tried that for a while but gave up. Every time the script is changed from upstream, you would need to update the patch.

MrByte commented:
> > Noob question. Can you make an ebuild for those of us that want to use it, ala local overlay?
> 
> Not really. You can make a user patch, but the script is too much of a moving target. I tried that for a while but gave up. Every time the script is changed from upstream, you would need to update the patch.

Yeah, sadly true. Right now I've got a script "zmerge" that looks for "world" in the parameters and does a ZFS snap for me if that's there.

I may take a swing at a patch at some point, or some other form of hackery.

Thanks for the reply.

floppym commented:
Hey, sorry for ignoring this PR. I'm not really on board with the idea in concept, so I'm not very motivated to re-review it. I think someone else will need to run with this one.

MrByte commented:
It gives options, options are always a good thing.

nagelp commented:
Just a user's opinion here: I'd really like to have the ability to go back in time after emerge has done changes to the system. This feature would enable a btrfs snapshot hook before emerge makes changes. It can make life much more convenient, and it's standard practice in other distributions already - so why not allow the user to do this on Gentoo as well?
